### PR TITLE
Refactoring `Enso_File` to be path based

### LINKED ISSUE #9289: Refactor `Enso_File` to be path-based
Currently `Enso_File` contains metadata like the asset `id`. This makes it only possible to refer to Enso files that already exist. As we also want to be able to have a `write` operation that may create new files, that is not enough. We need to be able to refer files that do not yet exist.

- [x] We should refactor `Enso_File` to be based on a path instead (much like `S3_File` works).
	- [x] Change the constructor to hold the `organization_name` and `path` (these are parallel to S3 `bucket` and `key`).
	- We should consider if we can re-use the S3 Path logic.
	- [x] Now `asset_type` needs to fetch. It should fail if the file currently does not exist.
	- [x] Update other methods to resolve the asset id based on the path and then access the API (e.g. size, read etc.)
- [x] Currently `/` throws `Common.Not_Found` if a file is not found. Now, it should just create a new path. 
	- [x] The error should be 'deferred' until the path is actually accessed, e.g. in `read`, `size`, 
		- [x] and it should be `File_Error.Not_Found` instead.
	- [x] Update tests to the new behaviour - check creating nonexistent paths with nonexistent parents.
- [x] We can now implement `Enso_File.parent` as it is now only path manipulation and does not need to perform any requests.
	- [x] Add tests for that as well.

Writing will be implemented as a separate task.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #9289 
- Ensures that we can refer through `Enso_File` to files that do not _yet_ exist - preparing us for implementing the Write functionalities for `Enso_File` (#9291).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso [hunk 1]
```diff
@@ -0,0 +1,103 @@
+private
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso:1]: +1

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso:1]: Great to see the `Internal` modules to also be marked as `private`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 2]
```diff
@@ -38,6 +38,7 @@ from project.Data.Text.Extensions import all
 from project.Enso_Cloud.Public_Utils import get_required_field
 from project.System.File_Format import Auto_Detect, Bytes, File_Format, Plain_Text_Format
 from project.System.File.Generic.File_Write_Strategy import generic_copy
+from project.System.File import find_extension_from_name
 
 type Enso_File
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: I was hoping to find here an explanation why there is `Enso_File`, when we already have `System.File`? But no, I don't see the reason.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: Fair, I will add some comment.

In very short - `Enso_File` is a reference to a file on the Enso Cloud, it is a sibling to e.g. `S3_File`.

Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: It seems to me as a copy of `System.File` created because you guys wanted to abstract something. However there already is [FileSystem abstraction](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/io/FileSystem.html) built into Truffle and it is fully accessible from `System.File`.

Shouldn't there just be an implementation of [org.graalvm.polyglot.io.FileSystem](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/io/FileSystem.html) when Enso engine is connected to Enso cloud?

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: We've been considering to do this at some point, but for the first iterations it seemed like an overkill. 

This PR is _not introducing _ `Enso_File` anyway, it is merely refactoring how it works internally - so if we want to migrate to the `FileSystem` based approach, let's create a separate ticket.

Still, I'm not 100% sure it is something we need right now, especially as while it may work OK with `Enso_File`, I'm not sure if it will work with the sibling `S3_File` - there we also hold credentials corresponding to a given S3 bucket access, and I'm not sure if the Truffle `FileSystem` API would allow us to do that?

**Comment by radeusgd** [comment]
Marking as ready for review.

I thought I'd consider some caching solution, but honestly we are not accessing the service more than in case of e.g. S3 files.

There is some opportunity for caching (i.e. we can cache some asset information from directory listing, currently we get that info, drop it and if asked about the file - re-fetch it), but it feels like a separate feature - so if we want to implement it - that should be done as a separate PR.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: Tracked in #9600

Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:43]: > hold credentials corresponding to a given S3 bucket access

An example of virtualization is a [memory-based file system](https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.api.test/src/com/oracle/truffle/api/test/polyglot/MemoryFileSystem.java) and it seems to do some [principal magic](https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.api.test/src/com/oracle/truffle/api/test/polyglot/MemoryFileSystem.java#L713).

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Public_Utils.enso [hunk 3]
```diff
@@ -24,7 +24,7 @@ get_required_field (key : Text) js_object (show_value : Boolean = True) (expecte
             suffix = if show_value then " in "+js_object.to_display_text else ""
             Error.throw (Enso_Cloud_Error.Invalid_Response_Payload "Missing required field `"+key+"`"+suffix+".")
         result = js_object.get key if_missing=handle_missing
-        if result.is_a expected_type then result else
+        result.if_not_error <| if result.is_a expected_type then result else
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Public_Utils.enso:27]: Why is this needed? 
Shouldn't `result.is_a` rethrow the error?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_User.enso [hunk 4]
```diff
@@ -47,10 +50,13 @@ type Enso_User
...
-    root_folder_id = get_required_field "rootDirectoryId" that expected_type=Text
-    root_folder = Enso_File.Value "" root_folder_id "" Enso_Asset_Type.Directory
-    is_enabled = get_optional_field "isEnabled" that if_missing=True expected_type=Boolean
+    id = get_required_field "userId" that expected_type=Text
     name = get_required_field "name" that expected_type=Text
     email = get_required_field "email" that expected_type=Text
-    id = get_required_field "id" that expected_type=Text
-    Enso_User.User name email id root_folder is_enabled
+    # Organization id is currently not being used
+    # organization_id = get_required_field "organizationId" that expected_type=Text
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_User.enso]: Just remove these?

**Review by jdunkerley — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Public_Utils.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Public_Utils.enso:27]: No, `is_a` does not propagate:
```
> (Error.throw "A")
>>> (Error: 'A')
> (Error.throw "A") . is_a Any
>>> true
> (Error.throw "A") . is_a Error
>>> true
> (Error.throw "A") . is_a Text
>>> false
> (Error.throw "A") . is_a Integer
>>> false
```

as you can see `is_a` 'swallows' the error and will return `true` for `x.is_a Error`.
We can change that behaviour, but until then, this `if_not_error` is necessary.
